### DISCUSSION TITLE
client connection drops and timeouts during rabbitmq_queue_messages_published_total statistic drop

### DISCUSSION THREAD

**Discussion by motmot80** [rabbitmq/rabbitmq-server#4743] [Other] [answered]
# Environment
RabbitMQ: 3.19.13
Erlang version: 24.3.2

# Error description
We are seeing peaks of send timeouts, connection drops and tls handshake timeouts sporadically but recurring for a short period of time. All cluster nodes within our cluster were affected independently from their load. 

# Reproduction steps
Unknown.

# Analysis
We analyzed this problem for several months not having a clue whats causing the issue except rabbitmq itself.

**BUT** today we found out that these events **EXACTLY** correlate with massive drops in the `rabbitmq_queue_messages_published_total` statistics. This seems to be another indication that the problem is located within the rabbitmq.

![grafik](https://user-images.githubusercontent.com/1308503/167017994-e2787170-066a-44ed-b7db-4b598771cad1.png)

We don't know if this is the root cause or just another aspect of the problem. All other rabbitmq statistics seem to be normal before, during and after the event. 

We don't even understand why the `rabbitmq_queue_messages_published_total` statistics could drop without deleting any queues. Garbage collection should be no issue in erlang. So what could cause these events?

@gerhard @michaelklishin Maybe you could give a hint what could cause such statistic drops, so we can dig deeper into the root cause.

Thanks in advance and best regards
Thomas

**Comment by michaelklishin** [comment, marked as answer]
Correlation does not mean causation. Metrics related to publishing will drop when a publishing rate drops, in particular, when publishing connections can no longer publish for any reason. No queues
have to be deleted for that to happen.

Both can easily have the same root cause: anything that prevents network activity on a connection
(could be a resource alarm or anything around the infrastructure involved) will affect both clients
and metric values reported.

The idea that a metric drop can lead to client connection closure seems a bit far fetched to me.
It very likely works the other way around.

**Reply by motmot80** [reply]
I really don't think that this is a causation. 

So - if I understand you right - you think that the dropped publishers are causing the statistics drop?

But the connection drops are not in the same dimension:

rabbitmq_channels:
![grafik](https://user-images.githubusercontent.com/1308503/167023760-83719a4e-da92-447d-8e50-571695895999.png)

rabbitmq_connections_closed_total:
![grafik](https://user-images.githubusercontent.com/1308503/167023952-841ccac3-2fa3-4453-8a20-c909e974249e.png)

rabbitmq_queue_messages_published_total:
![grafik](https://user-images.githubusercontent.com/1308503/167024246-ea880e4f-ff5e-43d0-aa47-4a6fadb70378.png)

**Reply by motmot80** [reply]
Concerning the network during this phase (1st cluster node):

Traffic:
![grafik](https://user-images.githubusercontent.com/1308503/167025205-77b0f85c-ffbc-4bf4-9c57-bbe044be1a1a.png)

Sockets:
![grafik](https://user-images.githubusercontent.com/1308503/167025287-d15caa66-d1d6-465c-a910-d21929ead343.png)

**Reply by michaelklishin** [reply]
Metric calculation is derivative from actual publisher activity and does not affect it. I don't see
how any parts of the code that compute, store or serve metrics can affect publishers.

I can think of plenty of scenarios where something affects new and currently open publisher
connections, and that also causes a rapid drop in publisher-related metrics.

In fact, in applications with *short lived* publishers that [churn through connections](https://www.rabbitmq.com/connections.html#high-connection-churn) a lot it will be visible particularly quickly.

**Reply by michaelklishin** [reply]
There is an uptick in the number of sockets used. If a RabbitMQ node, a proxy or something like that goes above the open file handle limit, no new connections will be accepted for a period of time.
With short lived publishers that will also result in a sharp drop of several publisher-related metrics.

These are the things I'd look out for, starting with potential evidence in the logs.

**Comment by motmot80** [comment]
I understand and follow your reasoning. No causation - just a correlation - which might help to find the root cause.
The connection churn is pretty much the same every day and is throttled to about 20 connections/sec on each cluster node.

So just to understand this in detail.

There's a drop of about 20 million published messages in the statistics (70% total). 
Having **all** connections restartet at night did not even once cause this high statistic drop rate.
So following your theory there must be a very small amount of publishers having produced round about 20 million messages in total disconnecting at ~ 8 AM the last two days but not the day before.

![grafik](https://user-images.githubusercontent.com/1308503/167035664-bac50c53-59a5-4548-952f-b1dedd6d86a6.png)
![grafik](https://user-images.githubusercontent.com/1308503/167035788-ccbc532b-98e2-40e1-b991-75d8e68e663c.png)
![grafik](https://user-images.githubusercontent.com/1308503/167035892-df06baf0-9308-43a0-bed3-8f972fe11c90.png)


So I tried to find the erlang code which is resetting the metrics:

```
rabbitmq_queue_messages_published_total ==
queue_messages_published_total ==
channel_stats(queue_exchange_stats, publish, Id, Value) ==
channel_queue_exchange_metrics (ETS Id 2)
```

Within the **channel** (isn't handle_consuming_queue_down_or_eol the consumer side?):
https://github.com/rabbitmq/rabbitmq-server/blob/cca1164afe1b39df4067de5d8ba74360cb0f010b/deps/rabbit/src/rabbit_channel.erl#L2443

And during a **metric garbage collection** event (every 2 minutes?)
https://github.com/rabbitmq/rabbitmq-server/blob/7abf749a60458aba9d6c9e6bdec1d3b0b2254007/deps/rabbit/src/rabbit_core_metrics_gc.erl#L97

Because we are having a huge cluster with several thousand queues and bindings the garbage collection caught my attention.
Could it be that the garbage collection todos are building up for longer than the 2 minute interval? 
Maybe the local queue gb_set is getting out of sync or the query for queues is timing out.

Taking a look at the ets allocations:

1st node (24h)
![grafik](https://user-images.githubusercontent.com/1308503/167038811-58a27224-baf8-49ef-b1d3-08960e5229f9.png)

2nd node (24h)
![grafik](https://user-images.githubusercontent.com/1308503/167039223-80dd015d-6680-4dd2-a60a-7c7a4257436a.png)

3rd node (24h)
![grafik](https://user-images.githubusercontent.com/1308503/167038389-e3c0e59a-395e-4946-9b96-fa8c008a02fc.png)

There seems to be a magical limit for the ETS memory of about 1.5 GB which is hit 30 minutes before the "statistic events" are starting. 
Just to mention there's plenty of RAM and CPU left.

But maybe just another correlating event.

**Comment by lhoguin** [comment]
Hello! Please provide the logs around the time of one of those events.

**Reply by motmot80** [reply]
We were able to reproduce a similar scenario in the test environment. The connection churn there is 100% in our hands so we'll try to investigate and isolate a log and a reproduction scenario.

@lhoguin  Is there any statistics to get the total active publisher count in a cluster? Consumer statistics are available, but how to monitor the amount of active publishers?

**Reply by michaelklishin** [reply]
FYI, we won't prioritize issues that require 100% connection churn. The docs pretty clearly say [it should be avoided](https://rabbitmq.com/connections.html#high-connection-churn). It is very difficult to reason about what resources may be temporarily exceeded or why and there isn't
much RabbitMQ nodes can do about the churn. Some resources involved are managed by the kernel.

**Reply by michaelklishin** [reply]
There is no such metric because there is no universally accepted definition of "an active publisher". The closest metric there is is connection egress data rate (outbound data rate) which for publishers will be constantly above a certain threshold. It won't be zero for consumers
either but on average, will be quite a bit lower.

**Reply by motmot80** [reply]
@michaelklishin I expressed myself misleadingly: "100% connection churn in our hands" means throttling the consumer attaches. The IoT production scenario is much more complex.

We have reproduced it in an isolated scenario.

# Reproduction steps:
1. Connect 100 publishers 
2. Constantly publish a small amount of messages (classic queues - non-persistent messages - 200 msg/sec per node)
3. Let the target queues fill up to about 1 million messages _(that's wasn't part of the normal daily test scenario)_
4. Start all the consumers (perftest) to consume the message backlog

# Expected:
Consumers are able to consume messages at a medium rate.
Publishers are able to send messages at a low rate.

# Actual:
Consumers are able to consume messages at about 2K msg/sec.
Publishers are timing out, causing them to close and reconnect.

The system load (1m) is at < 40%. 
The cpu is between < 50%.
The cluster normally is capable of about 15-20 K msg/sec.

On the other hand: 
Having almost no message backlog in the queues then the same consumer-attach (4) seems to have almost no impact on the publishers.

# Result:
@michaelklishin So what we see is **exactly what you predicted**. `rabbitmq_queue_messages_published_total` statistics are dropped when publishers are closing.
The root cause is that a huge amount of publishers were closed (send timeouts) although there were **plenty system ressources** left.
So the cluster wasn't capable of handling exchange publishes during a higher consuming rate. 

Unfortunately there is no metric getting the total publisher count. Maybe that would have helped us to find the cause faster.

Thanks very much for your support and keep up the good work.

We are working on a solution where the connection churn is reduced to about less than 10-20 connections/sec having more cluster nodes and less missbehaving devices.

Best regards
Thomas

**Reply by michaelklishin** [reply]
Thank you for reporting back.

I see how a metric that indicates the number of publishers would help. Unfortunately for us, the definition of a "publisher" will vary from user to user. It's a function of a publishing rate,
and what constitutes a "publisher" is a matter of opinion. Is 1 message an hour good enough to be considered a publisher? :)

But we will discuss how this can be done with a few folks who are involved with RabbitMQ monitoring.

**Reply by motmot80** [reply]
In my little clueless user/developer world I would have assumed exactly what the RabbitMQ docs are describing:

https://www.rabbitmq.com/publishers.html#terminology

*"[...]In general in messaging a **publisher** (also called "producer") is an application (or application instance) that publishes (produces) messages.[...]"*

And yes. In my personal opinion 1 message a year is good enough to be considered a publisher as long as the channel isn't closed! :D

**Reply by mkuratczyk** [reply]
@motmot80 Hey. I'd like to investigate that scenario you provided but I'm missing some of the details. Could you provide full perf-test commands to publish and consume? The exact numbers (number of queues, msg/s, etc) usually need to be adjusted to trigger a given issue on a certain hardware, so I'd appreciate if you could provide all other details so that I don't end up with something significantly different than what you have. 

Do you mean something like this?
```
# fill up the queues
perf-test -x 200 -y 0 -qa x-max-length=1000000 --auto-delete false -C 900000 -qp q-%d -qpf 1 -qpt 200 -c 1000

# start publishing slowly (here: 1 msg/s per queue)
perf-test -x 200 -y 0 -qa x-max-length=1000000  --auto-delete false -qp q-%d -qpf 1 -qpt 200 -c 1 -P 1

# start consuming while the previous command is running
perf-test -x 0 -y 200 -qa x-max-length=1000000 --auto-delete false -qp q-%d -qpf 1 -qpt 200
```

I'm not sure about the publisher confirms, consumer acks, consumption rate (`-R`), etc.
